docker/jenkins/Jenkinsfile

```diff
@@ -1,46 +1,858 @@
+  }
+}
+
+void cleanPreviousBuildFiles(String buildDir) {
+  // Delete any build files from a previous build
+  // Previous build folders affect findCopyZip() and can cause the stage to fail
+  sh "rm -rf ${buildDir}/*"
+}
+
+void createMultiMarkerScript(String markers, String testResultsFilename, String additionalOptions) {
+  // Passing multiple markers when running ./run-docker.sh does not work with bash.
+  // Therefore, create a script to maintain the single quotes that surround the markers
+  sh """echo "#!/bin/bash
+python -m pytest -m \'${markers}\' --junitxml=${testResultsFilename}.xml --html=${testResultsFilename}.html --self-contained-html ${additionalOptions}" >> run-tests.sh
+    """
+
+  // Give permissions to script
+  sh 'chmod 777 run-tests.sh'
+}
+
+void runDockerPytestWithMarker(String marker, String testResultsFilename, String additionalOptions) {
+  sh """./run-docker.sh python -m pytest -m ${marker} --junitxml=${testResultsFilename}.xml --html=${testResultsFilename}.html --self-contained-html ${additionalOptions}"""
+}
+
+void findBoardBuildFiles(String board, String searchDir, String dirToFind) {
+  def result = sh(script: "find $searchDir -type d -name \"$dirToFind*\"", returnStdout: true).trim()
+  if (result.empty) {
+      error "Directory containing '$dirToFind' not found."
+  }
+	return result
+}
+
+void findCopyZip(String board, String findDir, String copyDir, String stashName) {
+  def buildDir = findBoardBuildFiles(board, findDir, "hw_deployment_${board}")
+  sh "cp -r ${buildDir}/${board} ${copyDir}/"
+  dir(copyDir) {
+    sh "zip -r ${board}.zip ${board}/"
+    stash name: stashName, includes: "${board}.zip"
+  }
+}
+
+void createTestScript(String board, String marker, String testResultsFilename) {
+   if(board == "U250")
+    sh """echo "#!/bin/bash
+. /opt/xilinx/xrt/setup.sh
+. ${CONDA_ENV_ACTIVATE}
+python hack_driver_script.py
+python -m pytest -m ${marker} --junitxml=${testResultsFilename}.xml --html=${testResultsFilename}.html --self-contained-html" >> run-tests.sh
+    """
+  else
+    sh """echo "#!/bin/bash
+. /etc/profile.d/pynq_venv.sh
+. /etc/profile.d/xrt_setup.sh
+python hack_driver_script.py
+python -m pytest -m ${marker} --junitxml=${testResultsFilename}.xml --html=${testResultsFilename}.html --self-contained-html" >> run-tests.sh
+    """
+
+  // Give permissions to script
+  sh 'chmod 777 run-tests.sh'
+}
+
+void isNodeOnline(String labelName) {
+  Label label = Jenkins.instance.getLabel(labelName)
+  def agentOnline = false
+
+  if (label) {
+    List<Node> nodes = Jenkins.instance.getNodes()
+
+    nodes.each { node ->
+      if (node.getAssignedLabels().contains(label)) {
+        def computer = node.toComputer()
+        if (computer && computer.isOnline()) {
+          agentOnline = true
+        } else {
+          echo """Agent ${node.displayName} is offline"""
+        }
+      }
     }
+  } else {
+    echo """Node with label ${labelName} not found"""
+  }
+
+  return agentOnline
```

Comment:
This method returns a non null value, I think it would be better to declare with "def" (https://www.tutorialspoint.com/groovy/groovy_methods.htm) or with the return type.